Documentation/Coherency.md

```diff
@@ -0,0 +1,74 @@
+	-  If we don't turn on BuildAllConfigurations and BuildAllPackages, we will continue to have a large number of prebuilt packages.
+
+## Build multiple versions of each repo ##
+This would mean we build CLI with its expected version of CoreFX, and core-setup with its own expected version of CoreFX.
+Pros:
+- We match the official bits very closely.
+- We don't have to deal with servicing considerations.
+
+Cons:
+- This explodes rather quickly and would have extremely long build times.
+- We have to be careful about cross-contaminating package feeds during the builds.
+
+## Store and download packages in later builds ##
+During each build, we would save source-built packages off in an Azure account or NuGet feed.  During later builds, we would restore these packages and use them in builds later in the dependency tree.
+Pros:
+- Improved build times.
+	- Fringe benefit: incremental builds would probably finally work.
+- Still lets us eliminate prebuilt packages (all packages stored are source-built).
+- We match the official bits very closely.
+
+Cons:
+- Have to be very careful to not contaminate the storage with prebuilt packages.
+- Need internet access during the build.
+- Probably does not satisfy requirements for no prebuilt binaries.
+
+# Proposed Solution #
+---------------------
+Hybrid approach: we will enable storing and retrieving packages, but allow building multiple versions of each repo completely from source.  Daily builds and dev builds will mostly use the stored packages approach while release builds will use the completely-built-from-source approach.  Doing multiple versions of repos only for release builds should cut down on the numbers of builds that are done many times since these builds should be mostly coherent except for servicing considerations.
+
+## Phase 1 ##
+After Darc merges uberclone support (https://github.com/dotnet/arcade-services/pull/291), update our version of Darc and get rid of submodules, using an uberclone instead.  This will give us many more repos than we will actually use.  For the time being, use some `PreferredRepoHashes.props` or similar to tell the build which one to use - this should change nothing in our output.  At this point we should be able to take daily build updates but will be producing a product that differs from the official release, and will have problems if we don't implement later phases before servicing.
+
+## Phase 2 ##
+Implement some parts of repo isolation (https://github.com/dotnet/source-build/issues/928).  Use a separate output feed for each repo, then supply only the feeds that repo.projs say they depend on to each repo build.  Next implement publishing these feeds to our package storage.  It's nice to do this even if we don't have consumption yet because we will build up a cache of source-built packages we can use later.
```

Comment:
Seems like we can pack these up as RPM and Debian packages to address concerns about downloading and build times. I think it'd be reasonable to make them available as build outputs and distro maintainers can get those onto the N+1 build machines as dependencies.

(Not that this necessarily fits in Phase 2.)

Reply:
Added this to the store-and-download part of the spec.